internal/types/compare.go

```diff
@@ -40,12 +40,67 @@ const (
 )
 
 // Compare compares any BSON values in the same way as MongoDB does it for filtering.
+// When it compares scalar filterValue with array docValue, it returns first item
+// which is not Incomparable.
 //
 // It converts types as needed; that may result in different types being equal.
 // For that reason, it typically should not be used in tests.
 //
 // Compare and contrast with test helpers in testutil package.
 func Compare(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{greater: true, less: true, equal: true})
+}
+
+// CompareGreaterThan compares docValue for the greater than condition.
+func CompareGreaterThan(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{greater: true})
+}
+
+// CompareGreaterThanOrEq compares docValue for the greater than or equal to condition.
+func CompareGreaterThanOrEq(docValue, filterValue any) CompareResult {
```

Comment:
I think types package should export only `Compare` function and everything else should be an internal. I think we want to move towards a smaller exported APIs in general

Reply:
At the moment we pass `Compare(bson.A{"c", "b", "a"}, 43)` for `Compare(docValue, filterValue)`. So we don't have information related to the operator and we need that for this case. Perhaps we can pass `Compare(bson.A{"c", "b", "a"}, bson.D{{"$gte", 41.13}})` which includes the operator info, dependent on document comparison though.

Any other way of how we could achieve this?

Reply:
@chilagrow hmm, I think I start understanding... So, when we call `Compare` we can't say "it's equal" or "it's greater" without knowing the operator? Like for different operators, the result of `Compare` could be different?

Reply:
I think Compare(a, b) should compare values a and b; both a and b should not be expressions containing operators, but just values

Reply:
And I think `Compare(docValue, filterValue any)` is a rudiment of an old comparison code that tried to handle subarrays. That really should be `Compare(a, b any)` if possible

Reply:
I think what could possibly be done is to get `max` for  $gt, $gte and  `min` value for $lt,$lte before calling `Compare` and call `Compare(maxValue,filterValue)` for $gt for example. At the moment getting `max` or `min` is a bit tricky because when an array contains multiple types, it doesn't behave as expected.

Reply:
I'll see how this can be simplified to  Compare(a, b any)

Reply:
> At the moment getting max or min is a bit tricky because when an array contains multiple types, it doesn't behave as expected

So we just™ need to fix comparison, so Min and Max methods behave correctly

Reply:
I think I understand the logic of this PR and the solution looks reasonable to me. And yeah, I see a field to simplify our *existing* code, but I'm not sure if refactoring the code is in a scope of this issue.

Reply:
I moved all the logic related to operator into filter. No change in `Compare` now.